### PR TITLE
Save 'half timestep solutions' if using SolutionHistory with AdaptiveTimeSolver.

### PR DESCRIPTION
**Pull request by vikramvgarg**
If a two step adaptive time solver is used, we generate solutions at time + delta/2, which are now saved by Solution History and used for the adjoint calculations.

### CODE REVIEW

File: src/solvers/unsteady_solver.C [hunk 1]
```diff
@@ -197,6 +197,11 @@ void UnsteadySolver::adjoint_advance_timestep ()
       // The adjoint system has been solved. We need to store the adjoint solution and
       // load the primal solutions for the next time instance (t - delta_ti).
       _system.time -= _system.deltat;
+
+      if(_system.time < 0.0)
+      {
+        libmesh_error_msg("Error ! Time has become negative.");
```
Comment (roystgnr) [inline src/solvers/unsteady_solver.C:203]: Should this be a warning?  We typically start simulating at t=0, but that's a habit, not an intended requirement.  I could imagine someone simulating something (rocket startup?) where t=0 has a different mid-simulation meaning.  And since we're stepping back and forth by deltat here, don't we expect at least some numeric drift to be possible, even likely?

File: include/solvers/time_solver.h [hunk 2]
```diff
@@ -225,6 +225,19 @@ class TimeSolver : public ReferenceCountedObject<TimeSolver>
    */
   void set_solution_history(const SolutionHistory & _solution_history);
 
+  /**
+   * A setter function for the replace_stored_entry boolen owned by
+   * the solution history object
+   * */
+  void set_solution_history_replace_stored_entry(bool val)
+  { solution_history->set_replace_last_stored(val); }
```
Comment (roystgnr) [inline include/solvers/time_solver.h:233]: This seems way too specific, doesn't it?  Especially since we now have a getter, so get_solution_history()->set_replace_last_stored() is possible?

File: include/solvers/time_solver.h [hunk 3]
```diff
@@ -225,6 +225,19 @@ class TimeSolver : public ReferenceCountedObject<TimeSolver>
...
+
+  /**
+   * A getter function that returns a reference to the solution history
+   * object owned by TimeSolver
+   * */
+  std::shared_ptr<SolutionHistory> get_solution_history();
```
Comment (roystgnr) [inline include/solvers/time_solver.h:239]: So far we only need this for the outer time solver in the adaptive time stepping classes, right?  Would it make sense to make it protected at first, so we have a little more flexibility if we change our minds about the API in the future?

File: src/solvers/adaptive_time_solver.C [hunk 4]
```diff
@@ -64,6 +65,16 @@ void AdaptiveTimeSolver::init()
...
 
+  // Set the core_time_solver's solution history object to be the same one as
+  // that for the outer adaptive time solver
+  core_time_solver->set_solution_history(*(this->get_solution_history()));
+
+  // Now that we have set the SolutionHistory object for the coretimesolver,
+  // we set the SolutionHistory type for the timesolver to be NoSolutionHistory
+  // All storage and retrieval will be handled by the coretimesolver directly.
+  NoSolutionHistory outersolver_solution_history;
+  this->set_solution_history(outersolver_solution_history);
```
Comment (roystgnr) [inline src/solvers/adaptive_time_solver.C:76]: Since we're doing this reset anyway, plus retaining the use of SolutionHistory::clone(), do we still need to change solution_history to be a shared_ptr?  We didn't end up actually sharing it.

File: src/solvers/adaptive_time_solver.C [hunk 5]
```diff
@@ -87,27 +98,31 @@ void AdaptiveTimeSolver::reinit()
...
+  //     _system.time += _system.deltat;
+  //   }
+  //   else
+  //   {
+  //     // We are here because of a call to advance_timestep that happens
+  //     // via solve, the very first solve. All we are doing here is storing
+  //     // the initial condition. The actual solution computed via this solve
+  //     // will be stored when we call advance_timestep in the user's timestep loop
+  //     first_solve = false;
+  //   }
```
Comment (roystgnr) [inline src/solvers/adaptive_time_solver.C:116]: Maybe I read too much into his cleanup commits, but the copy of @jwpeterson who lives in my head is currently frowning at commented-out-but-undeleted code.

If we don't need this anymore (which surprises me, since I thought we'd still have to at least call core_time_solver->advance_timestep() from here; thanks for the comment below) then we should delete it entirely.

File: src/solvers/adaptive_time_solver.C [hunk 6]
```diff
@@ -121,7 +136,28 @@ void AdaptiveTimeSolver::advance_timestep ()
      _system.get_dof_map().get_send_list());
 }
 
+void AdaptiveTimeSolver::adjoint_advance_timestep ()
+{
+  // Call the core time solver's adjoint_advance_timestep
+  core_time_solver->adjoint_advance_timestep();
```
Comment (roystgnr) [inline src/solvers/adaptive_time_solver.C:142]: This makes the lack of a shim to ```core_time_solver->advance_timestep()``` above even more surprising; surely the operations here should be more symmetric?

File: include/solvers/memory_solution_history.h [hunk 7]
```diff
@@ -72,11 +72,12 @@ class MemorySolutionHistory : public SolutionHistory
   typedef std::map<std::string, std::unique_ptr<NumericVector<Number>>> map_type;
   typedef std::list<std::pair<Real, map_type>> list_type;
   typedef list_type::iterator stored_solutions_iterator;
+  //typedef list_type::reverse_iterator stored_solutions_reverse_iterator;
```
Comment (roystgnr) [inline include/solvers/memory_solution_history.h:75]: If we didn't end up needing this let's get rid of it.

File: src/solvers/unsteady_solver.C (hunk 1 above)
Comment (vikramvgarg) [reply src/solvers/unsteady_solver.C:203]: Agreed, will make this into a warning, and change the 0 to a f(TOLERANCE).

File: include/solvers/time_solver.h (hunk 2 above)
Comment (vikramvgarg) [reply include/solvers/time_solver.h:233]: Makes sense. Will remove the redundant setter.

File: examples/adjoints/adjoints_ex5/adjoints_ex5.C [hunk 8]
```diff
@@ -498,17 +498,20 @@ int main (int argc, char ** argv)
           // Solve the forward problem at time t, to obtain the solution at time t + dt
           system.solve();
 
+          // Advance to the next timestep in a transient problem
+          // Increments the time, stores the solution and localizes the old nonlinear solution
+          // for the non-adaptive time solver, does nothing if the time solver
+          // is adaptive
+          libMesh::out << "Advancing timestep" << std::endl << std::endl;
+          system.time_solver->advance_timestep();
```
Comment (roystgnr) [inline examples/adjoints/adjoints_ex5/adjoints_ex5.C:506]: Why the change in position here?  (other than that it saves us a redundant "+ dt" later)

File: examples/adjoints/adjoints_ex5/adjoints_ex5.C [hunk 9]
```diff
@@ -578,9 +581,10 @@ int main (int argc, char ** argv)
       // Now that the adjoint initial condition is set, we will start the
       // backwards in time adjoint integration
 
-      // For loop stepping backwards in time
-      for (unsigned int t_step=param.initial_timestep;
-           t_step != param.initial_timestep + param.n_timesteps; ++t_step)
+      // A backwards in time while loop
+      unsigned int t_step = 0;
+
+      while(system.time > 0)
```
Comment (roystgnr) [inline examples/adjoints/adjoints_ex5/adjoints_ex5.C:587]: This again seems to hardcode a special t=0.  Not nearly as bad in an app as in the library code ... but it does seem to break the initial_timestep restart support here, which I think ought to be workable even if it's not nearly as useful.

File: src/solvers/adaptive_time_solver.C (hunk 5 above)
Comment (vikramvgarg) [reply src/solvers/adaptive_time_solver.C:116]: Regarding the call to core_time_solver->advance_timestep() call, that isnt necessary since the two step time solver does it.
And we need the two step time solver to do the second advance_timestep to make sure we do all the SH store operations before we mess with the deltat.

There is the broader question of the system and user both being able to fiddle with the time, which I think will create problems. Is there a reason the advancing of the timestep cannot always be handled by the timesolver ?

File: examples/adjoints/adjoints_ex5/adjoints_ex5.C (hunk 8 above)
Comment (vikramvgarg) [reply examples/adjoints/adjoints_ex5/adjoints_ex5.C:506]: That redundant + dt was a bug, it was making me think we had reached a target time earlier than we were actually doing. Fixing this is what necessitated in the input file parameters.

File: examples/adjoints/adjoints_ex5/adjoints_ex5.C (hunk 9 above)
Comment (vikramvgarg) [reply examples/adjoints/adjoints_ex5/adjoints_ex5.C:587]: If we use a fixed time step solver, we have n timesteps, with the two step solver, we have 2*n timesteps. I didnt want to add an if conditional, but I guess we are already doing that to check the gold values.

File: src/solvers/adaptive_time_solver.C (hunk 5 above)
Comment (vikramvgarg) [reply src/solvers/adaptive_time_solver.C:116]: Regarding the commenting, wanted to run this by you before I deleted those lines. We should be clear to remove them now.

File: examples/adjoints/adjoints_ex5/general.in [hunk 10]
```diff
@@ -20,7 +20,7 @@ deltat = 0.1
 # Make this 0 to use a constant deltat
 timesolver_tolerance = 0.0
 timesolver_upper_tolerance = 0.0
-timesolver_maxgrowth = 1.09
+timesolver_maxgrowth = 1.12
```
Comment (roystgnr) [inline examples/adjoints/adjoints_ex5/general.in:23]: Why the change here?  I know we have to change the gold standard results anyway since we're doing the integration more accurately, but it'd be nice to avoid an apples-to-oranges difference if we can at least get apples-to-crabapples.

Comment (vikramvgarg) [reply examples/adjoints/adjoints_ex5/general.in:23]: See above, the extra dt was making me think we reached 1.00 when we had only reached 0.9, so kept the number of timesteps the same, increased the size of the steps we are taking.

File: examples/adjoints/adjoints_ex5/adjoints_ex5.C (hunk 9 above)
Comment (roystgnr) [reply examples/adjoints/adjoints_ex5/adjoints_ex5.C:587]: But we don't have 2*n timesteps.  We have n timesteps, with 2*n sub-timesteps.  TwoStepTimeSolver::adjoint_solve should be handling the 2 core_time_solver->adjoint_solve() calls, the same way TwoStepTimeSolver::solve handles 2 core_time_solver->solve() calls...

File: src/solvers/adaptive_time_solver.C (hunk 5 above)
Comment (roystgnr) [reply src/solvers/adaptive_time_solver.C:116]: Yeah, I think ideally advancing the time would only be done by advance_timestep().  Isn't that how it works now?  I don't see system.time being touched outside of advance_timestep().

File: examples/adjoints/adjoints_ex5/general.in (hunk 10 above)
Comment (roystgnr) [reply examples/adjoints/adjoints_ex5/general.in:23]: Okay.  So can this be reverted then?

File: examples/adjoints/adjoints_ex5/adjoints_ex5.C (hunk 9 above)
Comment (vikramvgarg) [reply examples/adjoints/adjoints_ex5/adjoints_ex5.C:587]: This would mean changes to the adjoint_solve() method of system, it will have to call a different adjoint_solve() implementation depending on the type of timesolver we are using. Do we want those changes to be part of this PR ?

File: src/solvers/adaptive_time_solver.C (hunk 5 above)
Comment (jwpeterson) [reply src/solvers/adaptive_time_solver.C:116]: The copy of me living in @roystgnr's head:

ralph

Comment (roystgnr) [reply src/solvers/adaptive_time_solver.C:116]: Sorry, GitHub hasn't been updating pages on me properly.  Definitely clear to remove.

File: examples/adjoints/adjoints_ex5/adjoints_ex5.C (hunk 9 above)
Comment (roystgnr) [reply examples/adjoints/adjoints_ex5/adjoints_ex5.C:587]: It wouldn't be any change to ImplicitSystem::adjoint_solve(), it would be a change to DifferentiableSystem::adjoint_solve() and an addition of TimeSolver::adjoint_solve() .. which in hindsight we should have had years ago.  Instead of just setting a flag before handing off to ImplicitSystem, DifferentiableSystem would be calling time_solver->adjoint_solve(), analogous to what it does in the solve() case.  Some of the TimeSolver subclasses would still be able to just set a flag, but things like TwoStep would actually do two steps.

**Comment by vikramvgarg** [comment]
I think its better if I cancel this PR, reorder and break things down a bit. Plan is for a sequence of three branches:

1) Change signature of retrieve and store to take time as an argument.
2) Generalize adjoint_solve.
3) Full support for two step time solver (SolutionHistory::erase(Real time) and other changes to TwoStepTimeSolver).

**Comment by roystgnr** [comment]
Thanks!  Sorry I underestimated how far we were from doing adjoint_solve() properly.
